🎉 feat: first commit

Jpgomes06 opened the pull request:
### Descrição

Este PR adiciona a funcionalidade de criação de contas TOTP via rota HTTP (**POST /2fa/totp**).  
Quando a rota é chamada, o sistema gera:

- um id (UUID)
- um secret
- a URL padrão para autenticação (**otpauth://…**)
- o QR Code em formato **base64**

Depois disso, essas informações são enviadas para a fila RabbitMQ (otp-accounts).

O fluxo funciona assim:
1. O cliente faz a requisição HTTP
2. O serviço gera UUID, secret e QR Code
3. Envia para a fila RabbitMQ
4. retornando por fim essas informações

---

#### 📌 envio das informações do usuário  
<img width="1228" height="290" alt="Screenshot from 2025-12-03 22-14-13" src="https://github.com/user-attachments/assets/c2d807f8-20d1-4376-a605-2ad57a77b317" />

#### 📌 Dashboard do RabbitMQ  
<img width="1305" height="210" alt="image" src="https://github.com/user-attachments/assets/7123389c-8229-4369-9bf9-55b5fa6970f3" />

#### 📌 Retorno no postman 
<img width="1254" height="422" alt="image" src="https://github.com/user-attachments/assets/0f51f492-3df8-4da7-a49d-f38da7330d57" />

Jpgomes06 committed:
🎉 feat: first commit

Jpgomes06 committed:
📚 docs: update Reame.md

Jpgomes06 committed:
🎉 feat:add utc and secret base32

Jpgomes06 committed:
🎉 feat:add qrcode function

Jpgomes06 committed:
🎉 feat:add rabbitmq gateway

Jpgomes06 committed:
🎉 feat:add config files

Jpgomes06 committed:
🎉 feat:add producer and error handler

Jpgomes06 committed:
♻️ refactor: start implementing DIP

Jpgomes06 committed:
♻️ refactor: refactor create totp feature with solid

Jpgomes06 committed:
♻️ refactor: Moving the createTotpProvider file to the repositories folder.

Jpgomes06 committed:
♻️ refactor: renamed createTotpProvider

Jpgomes06 committed:
🎉 feat: adding unit tests and jest config.

Luiz-Cruz commented on src/controllers/createAccountController.ts:
```diff
@@ -0,0 +1,19 @@
+            data: {
+                totp: totpCreated
+            }
+        })
```
Sugestão: Evite hardcode de status HTTP. Use as constantes da biblioteca http-status-codes, como StatusCodes.OK no lugar de 200. Isso melhora a legibilidade e padroniza o código.

Jpgomes06 replied:
feito

Luiz-Cruz commented on src/infra/rabbitmq-gateway.ts:
```diff
@@ -0,0 +1,33 @@
+            this.connection = await connect(process.env.RABBITMQ_URL || 'amqp://rabbitmq:5672');
+            this.channel = await this.connection.createChannel();
+            await this.channel.assertQueue(this.queueName, { durable: true });
+            console.log('Connected to RabbitMQ');
```
Sobre o log "Connected to RabbitMQ", acho melhor remover. As boas práticas de logging recomendam focar em registrar apenas erros e eventos realmente relevantes. Logar operações de sucesso ou de rotina, como uma conexão feita com sucesso, só vai poluir os logs e dificultar a análise quando um problema real acontecer. Se a conexão falhar, aí sim faz sentido logar o erro com os detalhes para o diagnóstico. Essa abordagem é bem comum em ambientes produtivos e está em guias como este aqui:

https://betterstack.com/community/guides/logging/nodejs-logging-best-practices/?utm_source=openai

Jpgomes06 replied:
feito

Luiz-Cruz commented on src/infra/rabbitmq-gateway.ts:
```diff
@@ -0,0 +1,33 @@
+    private connection?: ChannelModel;
+    private channel?: Channel;
+    private readonly queueName = process.env.QUEUE_NAME || 'otp-accounts';
+
+    public async connect():Promise<void> {
+        try {
+            this.connection = await connect(process.env.RABBITMQ_URL || 'amqp://rabbitmq:5672');
+            this.channel = await this.connection.createChannel();
+            await this.channel.assertQueue(this.queueName, { durable: true });
+            console.log('Connected to RabbitMQ');
+        } catch (error) {
+            throw new Error('Failed to connect to RabbitMQ.');
+        }
+    }
+
+    public async publish(totpAccount: TotpAccount): Promise<boolean | undefined> {
+        try {
+            if (!this.channel) {
+                await this.connect();
+            }
+            const message = JSON.stringify({totpAccount});
+            return this.channel?.sendToQueue(this.queueName, Buffer.from(message), {
+                persistent: true,
+            })
+        } catch (error) {
+            throw new Error('Internal server error.');
+        }
+    }
+}
```
Nos catches, ao lançar novos erros com mensagens genéricas, a mensagem original e o stack trace do erro real acabam se perdendo, o que dificulta bastante o diagnóstico. O ideal é sempre preservar o contexto original do erro. Como já estamos usando a versão mais recente do Node.js, podemos (e devemos) usar o parâmetro `cause` no construtor do `Error` para encadear o erro original — assim, você tem acesso fácil à origem do problema, inclusive ao stack trace, de acordo com a [documentação oficial do Node.js](https://nodejs.org/api/errors.html#errorcause):

**Exemplo recomendado:**
```typescript
catch (error) {
  throw new Error('Failed to connect to RabbitMQ', { cause: error });
}
```
Com isso, se for preciso capturar esse erro em outro lugar, você consegue acessar tanto a mensagem customizada (`err.message`) quanto o erro original (`err.cause` e o stack original).

**Exemplo de acesso depois:**
```typescript
// Em ambiente de produção, o ideal é usar um logger ao invés de console.error:
logger.error(err.message);
logger.error(err.cause);  // ou logger.error(err) para registrar tudo
```
Esses exemplos usam `console.error` só para demonstrar o acesso aos campos, mas o correto mesmo é usar o logger da aplicação, pois assim você centraliza e padroniza os logs.

Assim, o log do erro fica mais completo e realmente útil para análise e correção.

Luiz-Cruz commented on src/middlewares/error.ts:
```diff
@@ -0,0 +1,6 @@
+import { NextFunction , Request , Response } from  "express" ;
+
+export  const  errorHandler = ( err: Error , req: Request, res: Response, next: NextFunction ) => {
+    console.error (err);
```
O mesmo aqui referente aos logs

Luiz-Cruz commented on src/middlewares/error.ts:
```diff
@@ -0,0 +1,6 @@
+
+export  const  errorHandler = ( err: Error , req: Request, res: Response, next: NextFunction ) => {
+    console.error (err);
+    res.status ( 500 ). send ({ errors : [{ message : err.message }] });
```
Nesse caso seria melhor não retornar o err.message, pois pode expor o erro diretamente pela rota da sua API, acho interessante colocar uma mensagem genérica por enquanto, como: "Internal server error. Please try again later."

Sugiro também adicionar o status com o mesmo padrão que sugeri no controller.


e o send algo como:

```
message: "Internal server error. Please try again later"
```

Sem o "errors"...

Luiz-Cruz commented on src/index.ts:
```diff
@@ -0,0 +1,28 @@
+
+app.listen(PORT,() => {
+    console.log(`Server running on http://localhost:${PORT}`);
+});
```
adicionar o padrão de log que comentei, mas do tipo "info"

Luiz-Cruz commented on Reame.md:
Sugiro apagar o arquivo, como não há nenhuma informação nele

Luiz-Cruz commented on solid.ts:
```diff
@@ -0,0 +1,83 @@
+interface IPaymentMethod {
```
creio que subiu esse arquivo sem querer

Luiz-Cruz reviewed (commented):
Deixei comentários acima

Luiz-Cruz reviewed (changes requested):
Comentários acima

Luiz-Cruz commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+/node_modules
+.idea
+.env
```
faltou agregar a pasta que o Typescript gera ao compilar, não?

Jpgomes06 committed:
♻️ refactor:

- Adding the following libraries: http-status-codes for non-hardcoded status codes and pino to use the logger

- Adding /dist to the .gitignore

- Changing the API initialization message and errors passed through the middleware

Luiz-Cruz commented on src/middlewares/error.ts:
```diff
@@ -0,0 +1,5 @@
+import { NextFunction , Request , Response } from  "express" ;
+
+export  const  errorHandler = ( err: Error , req: Request, res: Response, next: NextFunction ) => {
+    res.status ( 500 ). send({ message : "Internal server error. Please try again later"});
```
Colocar o status vindo da lib de status codes

Luiz-Cruz commented on src/index.ts:
```diff
@@ -0,0 +1,33 @@
+
+const app = express();
+const PORT = 3000;
+const logger = pino();
```
Top!

Luiz-Cruz commented on src/infra/rabbitmq-gateway.ts:
```diff
@@ -0,0 +1,34 @@
+            this.connection = await connect(process.env.RABBITMQ_URL || 'amqp://localhost:5672');
+            this.channel = await this.connection.createChannel();
+            await this.channel.assertQueue(this.queueName, { durable: true });
+        } catch (error) {
```
nos cenários de erro, vc pode logar o erro com o pino, assim como fez no inicio do app

Luiz-Cruz reviewed (changes requested):
Usar logs no catch dos erros.( log de erro)

Jpgomes06 committed:
♻️ refactor:

- Changing the 500 status to StatusCodes.INTERNAL_SERVER_ERROR in the error file

- Adding a logger to the catch blocks in the rabbitmq-gateway file
